ateam_ai/src/plays/defense.hpp

```diff
@@ -0,0 +1,125 @@
+// Copyright 2023 A Team
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+
+#ifndef PLAYS__DEFENSE_HPP_
+#define PLAYS__DEFENSE_HPP_
+
+#include <Eigen/Dense>
+#include <array>
+#include <vector>
+#include <optional>
+
+#include "types/world.hpp"
+#include "types/behavior_goal.hpp"
+#include "ateam_geometry/utilities.hpp"
+#include "ateam_geometry/ateam_geometry.hpp"
+
+const FieldSidedInfo & our_side_info
+DirectedGraph<BehaviorGoal> generate_basic_defense(
+  const World & world,
+  const Feild & field,
+  const FieldSidedInfo & our_side_info)
+{
+  DirectedGraph<BehaviorGoal> defense_graph;
+  // Go to the middle of the goalie area
+  BehaviorGoal goalie = get_goalie_behavior_goal(our_side_info);
+  defense_graph.add_node(goalie);
+
+  std::vector<BehaviorGoal> defenders = get_defense_behavior_goals(world, field, 2);
+  for (BehaviorGoal defender : defenders) {
+    defense_graph.add_node(defender);
+  }
+  return defense_graph;
+}
+
+BehaviorGoal get_goalie_behavior_goal(const FieldSidedInfo & our_side_info)
+{
+  // Line that is 0.5 m from the defense area in all directions
+  ateam_geometry::Segment goalie_line = ateam_geometry::Segment(
+    ateam_geometry::Point(
+      -4,
+      0.5), ateam_geometry::Point(
+      -4, -0.5));
+
+  // Get the ball location
+  std::optional<Eigen::Vector2d> ball_location = world.get_unique_ball();
+  while (!ball_location.has_value()) {
+    ball_location = world.get_unique_ball();
+  }
```

Comment:
Get unique ball should either be the actual ball or nothing. Nothing means there is no ball. I think this would be an infinite loops if we have no ball.

Reply:
I think there are given situations where we may not have a ball yet (or not have filtered one yet) but may want to call a play anyway... i.e., if the referee is moving the ball and it can't be seen properly by vision. So maybe we can call this a limited number of times (say 10ish max?) and then if it fails too much, we return an empty vector?

Reply:
I think that's the best we can do. Especially while the trajectory follower controller is still in the ai module. Once that is removed, we can get away with more weird cycle times